1-Draft/RFC00XX-Module-Isolation.md

```diff
@@ -0,0 +1,251 @@
+
+So, I propose to add the new internal property `ExecutionContext.ActiveReflectionContext`,
+which points to the currently active ALC that is supposed to be visible and usable in PowerShell.
+If its value is `null`, then the currently active ALC is the default ALC.
+
+Similarly, the setup and restore of this property should be done before and after command invocation:
+- `ExecutionContext.ActiveReflectionContext` is set before invoking a module command,
+  and restored to the previous context afterwards.
+- `ExecutionContext.ActiveReflectionContext` is set before running a command or script block in a module context (`& $module <cmd-or-script>`),
+  and restored to the previous context after the invocation.
+
+### Update Type Resolution Logic
+
+Now that we have custom ALC's for modules, the type resolving logic needs to be changed:
+
+1. We should first look for the type within the `ExecutionContext.ActiveReflectionContext`,
+and then revert back to the default ALC.
+1. When reverting back to the default ALC,
+we should ignore the assemblies that are loaded (and thus already searched) from the active ALC.
+
+Since different versions of the same assemblies can be loaded into a custom ALC in PowerShell,
+we may have type identity issues within a session -- type `C` and `C'` are the same type in terms of the fully-qualified type name,
+but they are from two different assembly instances that are loaded into two load contexts.
+
+In order to avoid type identity issue in type resolution and have it work consistently,
+we need to do the following extra updates:
+
+1. Today, we have a `TypeCache` data structure that maps a `Tuple<ITypeName, TypeResolutionState>` to a `Type`.
+We will need to have separate `TypeCache` defined in the `PSModuleLoadContext` class,
+so every custom ALC instance has its own type cache.
+The existing `TypeCache` will become the type cache for the default ALC.
+
+1. Today, we have `ExecutionContext.AssemblyCache` that holds all assemblies that are explicitly loaded by PowerShell.
+Simiarly, we will need to have separate `AssemblyCache` defined in the `PSModuleLoadContext` class,
+so every custom ALC instance has its own `AssemblyCache`.
+The existing `AssemblyCache` will be for the default ALC exclusively.
+
+1. All the `ITypeName` implementation types supports type caching.
+The cached type of a `ITypeName` instance should be able to invalid the caching when `ExecutionContext.ActiveReflectionContext` changes.
+
+## Potential Problems
+
+### Module Interoperability
+
+Imagine the assemblies A and B are loaded into ALC-M (isolated module M).
+Assembly A exposes a cmdlet `Demo-Cmdlet` that has a parameter accepting an object of type `B::C`.
+
+Scenario 1:
+Let's say the default ALC doesn't contains the assembly B.
+Then PowerShell won't be able to resolve the type `[B::C]` from outside the module M (e.g. from global scope),
+because `[B::C]` can only be resolved within the ALC-M.
+
```

Comment:
My assumption is that if a conflicting assembly is attempted to be loaded into a new ALC, we will still see the same error that we see now. Is that right?

Reply:
@JamesWTruher The error we see today is that the conflicting assembly cannot be loaded, so we won't see this error when introducing new ALC. But I'm not sure if I understand your question correctly ... can you please elaborate a bit more on the question?